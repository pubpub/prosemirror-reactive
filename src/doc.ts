import { Node } from "prosemirror-model";

import { DEFAULT_ID_ATTR_KEY } from "./constants";
import { addTemporaryIdsToDoc, mapDoc } from "./util";
import { DocumentStore } from "./store/documentStore";

interface GetReactedDocOptions {
    missingIdGenerator?: () => string;
    idAttrKey?: string;
    documentState?: Record<any, any>;
}

/**
 * Get a new document with all of the reactive attrs of a source document computed in a single run.
 * @param doc
 * @param schema
 * @param options
 */
export const getReactedDoc = (doc: Node, options: GetReactedDocOptions = {}): Node => {
    const { idAttrKey = DEFAULT_ID_ATTR_KEY, missingIdGenerator, documentState = {} } = options;
    const docWithIds = addTemporaryIdsToDoc(doc, idAttrKey, missingIdGenerator);
    const store = new DocumentStore({
        nodeSpecs: doc.type.schema.nodes,
        invalidateNode: () => {},
        idAttrKey,
        documentState,
    });
    const invalidatedIds = Object.keys(store.run(docWithIds));
    return mapDoc(docWithIds, (node: Node) => {
        const id = node.attrs && node.attrs[idAttrKey];
        if (invalidatedIds.includes(id)) {
            const copy = store.getReactedCopy(node);
            if (copy) {
                return copy;
            }
        }
        return node;
    });
};
