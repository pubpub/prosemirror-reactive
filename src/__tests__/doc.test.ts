/* global describe, it, expect */
import { Node } from "prosemirror-model";

import { sheepDoc as sheepJson, counterAndHeaderDoc } from "../examples/docs";
import { createSchema, sheep, sheepNamer, counter, header } from "../examples/schemas";
import { getReactedDoc } from "../doc";

const schema = createSchema({ counter, header });
const doc = Node.fromJSON(schema, counterAndHeaderDoc);

const sheepSchema = createSchema({ sheep, sheepNamer });
const sheepDoc = Node.fromJSON(sheepSchema, sheepJson);

const getIdGenerator = () => {
    let id = 0;
    return () => {
        id++;
        return id.toString();
    };
};

describe("getReactedDoc", () => {
    const idGenerator = getIdGenerator();

    it("computes a reacted doc", () => {
        const reactedDoc = getReactedDoc(sheepDoc, { missingIdGenerator: idGenerator });
        expect(reactedDoc).toMatchSnapshot();
    });

    it("computes another reacted doc", () => {
        const reactedDoc = getReactedDoc(doc, { missingIdGenerator: idGenerator });
        expect(reactedDoc).toMatchSnapshot();
    });
});
