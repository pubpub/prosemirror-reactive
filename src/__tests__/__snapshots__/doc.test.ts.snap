// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getReactedDoc computes a reacted doc 1`] = `
Object {
  "content": Array [
    Object {
      "content": Array [
        Object {
          "attrs": Object {
            "id": "1",
            "report": "My sheep is named Shelly",
            "sheepId": "sheep-3",
          },
          "type": "sheepNamer",
        },
      ],
      "type": "paragraph",
    },
    Object {
      "content": Array [
        Object {
          "attrs": Object {
            "id": "sheep-1",
            "name": "Shaun",
          },
          "type": "sheep",
        },
      ],
      "type": "paragraph",
    },
    Object {
      "content": Array [
        Object {
          "attrs": Object {
            "id": "2",
            "report": "My sheep is named Seamus",
            "sheepId": "sheep-2",
          },
          "type": "sheepNamer",
        },
      ],
      "type": "paragraph",
    },
    Object {
      "content": Array [
        Object {
          "attrs": Object {
            "id": "3",
            "report": "I don't see my sheep",
            "sheepId": "sheep-4",
          },
          "type": "sheepNamer",
        },
      ],
      "type": "paragraph",
    },
    Object {
      "content": Array [
        Object {
          "attrs": Object {
            "id": "sheep-2",
            "name": "Seamus",
          },
          "type": "sheep",
        },
      ],
      "type": "paragraph",
    },
    Object {
      "content": Array [
        Object {
          "attrs": Object {
            "id": "sheep-3",
            "name": "Shelly",
          },
          "type": "sheep",
        },
      ],
      "type": "paragraph",
    },
    Object {
      "content": Array [
        Object {
          "attrs": Object {
            "id": "4",
            "report": "My sheep is named Shaun",
            "sheepId": "sheep-1",
          },
          "type": "sheepNamer",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;

exports[`getReactedDoc computes another reacted doc 1`] = `
Object {
  "content": Array [
    Object {
      "attrs": Object {
        "id": "5",
      },
      "content": Array [
        Object {
          "text": "Welcome to my document",
          "type": "text",
        },
      ],
      "type": "header",
    },
    Object {
      "content": Array [
        Object {
          "text": "Here is a bunch of stuff that isn't true",
          "type": "text",
        },
        Object {
          "attrs": Object {
            "count": 1,
            "id": "6",
          },
          "type": "counter",
        },
        Object {
          "text": "But we do have citations for it!",
          "type": "text",
        },
        Object {
          "attrs": Object {
            "count": 2,
            "id": "7",
          },
          "type": "counter",
        },
      ],
      "type": "paragraph",
    },
    Object {
      "attrs": Object {
        "id": "8",
      },
      "content": Array [
        Object {
          "text": "More stuff",
          "type": "text",
        },
      ],
      "type": "header",
    },
    Object {
      "content": Array [
        Object {
          "text": "Here is some more stuff",
          "type": "text",
        },
        Object {
          "attrs": Object {
            "count": 3,
            "id": "9",
          },
          "type": "counter",
        },
      ],
      "type": "paragraph",
    },
  ],
  "type": "doc",
}
`;
